monitor/monitor.go

```diff
@@ -68,7 +59,21 @@ func (m *Monitor) Run(npages int) {
 		}
 	}(t)
 
-	// Main event loop
+	eventConfig.NumPages = npages
+
+main_event_loop:
+	for ok := <-waitForListeners; !ok; {
```

Comment:
What is the purpose for this channel?

Reply:
To wait for listeners which are added in the handle connection. Am I using it wrong?

Reply:
There are a few patterns you might use that would be more canonical. The most useful would be that you can make a buffered channel (`make(chan bool, 1000)`) and then the writes you have to it below wouldn't need the goroutine wrappers. Since you're using `main_event_loop` as a normal label, instead of one for the loop, you can replace the for `ok := <-waitForListeners; !ok;` with a simple for `<-waitForListeners` but then only write to the channel when you would have written a true. You currently always write true in `handleConnection` so that doesn't change, and you would have to use an if statement in the `send` case.

Reply:
Looks cleaner. The only other change I did was drop the loop because it seemed to lock forever. Please take one more pass in case I misunderstood you.

Reply:
Or maybe wait looks like I missed something.

Reply:
Sort of. My goal was to avoid the busy loop on the `waitForListeners` channel. The thing is, we're mixing concerns a bit, in that we want the loop to pause until there are listeners but the only way we have to do that is by blocking on the channel every iteration, thus, we also need to bypass this channel read :/

Would it be simpler to coordinate this a different way?
- Spawn a goroutine that runs the logic in main_event_loop when we handleConnection and we go from numListeners==0 to numListeners==1
- In this main_event_loop goroutine, exit it completely when there are no more listeners 
after purgeInactiveListeners (or just check at the end/start of the loop).
- We can also defer closing sockets and things in this goroutine, so we know the cleanups happen at the right time.
- This version would not have this waitForListeners channel at all.

My reasoning is that it is simple to not run the perf reading parts at all if there are no listeners. I think this was the motivation for these changes right?

If you want to stick with what you have now, then it might be easier to split the gotos and labels:
- add a new label waiting_for_listers. jump to it when we have no more listeners (so, where we currently jump to main_event_loop)
- main_event_loop is moved to after the waitForListeners check. We jump to it in the case where there are listeners still, and we need to poll again and keep reading.

Reply:
You are right. Will do it as purposed, just need to test it a little bit locally and will push it out. Thanks, Ray.